packages/ts-transformers/src/closures/transformer.ts

```
@@ -0,0 +1,646 @@
+      return a.text === b.text;
+    }
+    return false;
+  }
+
+  // Replace captured expressions with their parameter names
+  for (const [varName, capturedExpr] of captures) {
+    const visitor: ts.Visitor = (node) => {
+      // Check if this node matches the captured expression
+      if (ts.isExpression(node) && expressionsMatch(node, capturedExpr)) {
+        return factory.createIdentifier(varName);
+      }
+      return ts.visitEachChild(node, visitor, undefined);
+    };
+    transformedBody = ts.visitNode(
+      transformedBody,
+      visitor,
+    ) as typeof transformedBody;
+  }
+
+  // Create the new callback
+  const newCallback = factory.createArrowFunction(
+    callback.modifiers,
+    callback.typeParameters,
+    [destructuredParam],
+    callback.type,
+    ts.isArrowFunction(callback)
+      ? callback.equalsGreaterThanToken
+      : factory.createToken(ts.SyntaxKind.EqualsGreaterThanToken),
+    transformedBody,
+  );
+
+  // Wrap in recipe() using the proper imported identifier
+  const recipeIdentifier = getHelperIdentifier(
+    factory,
+    context.sourceFile,
+    "recipe",
+  );
+  const recipeCall = factory.createCallExpression(
+    recipeIdentifier,
+    undefined,
+    [
+      factory.createStringLiteral("map with pattern including captures"),
+      newCallback,
+    ],
+  );
+
+  // Create the params object
+  const paramsObject = factory.createObjectLiteralExpression(paramProperties);
+
+  // Create map_with_pattern property access (e.g., state.items.map_with_pattern)
+  // mapCall.expression is a PropertyAccessExpression like state.items.map
+  // We need to replace "map" with "map_with_pattern"
+  const mapWithPatternAccess = ts.isPropertyAccessExpression(mapCall.expression)
+    ? factory.createPropertyAccessExpression(
+      mapCall.expression.expression, // state.items
+      factory.createIdentifier("map_with_pattern"),
+    )
+    : factory.createIdentifier("map_with_pattern"); // Fallback (shouldn't happen)
```

Let's throw an error if we don't expect this